### ISSUE TITLE
Perform pre-evaluation on all FM, to complete initialization of disc-dependent structures

### ISSUE THREAD

**Issue by bartgol** [sandialabs/Albany#856]
This may become more of an issue with panzer's DOF manager. Right now, some disc-related info is computed on the fly during evaluateFields (possibly cached, if its calculation is non trivial). It might be cleaner to do things in either preEvaluate or, even better, in postRegistrationSetup. Currently, we don't do it pretty much ever, but this PHX phase should be exploited more.

For instance, with the DOF manager, dofs offsets within a cell are the same on all cells, but require the disc object to be retrieved. This could be done in postRegistrationSetup. OTOH, offsets of dofs on the side require to know the side rel position inside the cell, which requires a Side object, which is only available at eval time (though we could grab the disc, extract the sideset (as a whole), and figure out the side_pos inside the cell for sides on this sideset).

Long story short, we could make the code cleaner by more extensively using evaluators methods that are called *before* evaluateFields.
